add tracing to langchain

agola11 opened the pull request:
* add implementations of `BaseCallbackHandler` that log traces to a locally running `langchain-server`

agola11 committed:
add tracing interface

agola11 committed:
add stdout impl and add calls

agola11 committed:
add examples, start implementing nestedjson

agola11 committed:
add example

agola11 committed:
add nested json

agola11 committed:
add basic tracing

agola11 committed:
delete redundant file

hwchase17 commented on langchain/llms/base.py:
```diff
@@ -48,7 +49,12 @@ def generate(
-            return self._generate(prompts, stop=stop)
+            get_tracer().start_llm_trace({"name": self.__class__.__name__}, prompts)
+            output = self._generate(prompts, stop=stop)
+            get_tracer().end_llm_trace(
```
why not just pass output in?

agola11 replied:
output is of type `LLMResult`... I know we talked about the base tracer taking in `LLMResult` as well, but that would result in a circular import. Even placing `LLMResult` in a separate file within `llms` would cause this problem bc of the imports within `__init__.py`

hwchase17 replied:
im down to put it in https://github.com/hwchase17/langchain/blob/master/langchain/schema.py, Generation is already there

hwchase17 commented on langchain/tracing/__init__.py:
```diff
@@ -0,0 +1,21 @@
+from langchain.tracing.stdout import StdOutTracer
+
+
+def get_tracer() -> BaseTracer:
```
why this vs just setting `langchain.tracer = ...`

when we have more complex tracers that could be initialized with api keys, etc, would be nice to let them linitialize it however and then just do `langchain.tracer = ...`

that would also allows users to easily implement and use their own tracers (which i guess could be good or bad)

hwchase17 replied:
is this a technical decision or a strategic one to do it this way?

agola11 replied:
I don't have strong opinions about this, this is just the way I chose to implement it atm.

I do think calling a function is a little bit cleaner than having to set a global variable, plus being able to switch tracing implementations by swapping environment variables rather than making code changes is nice too

hwchase17 replied:
maybe then we should do something at the top of the file like `tracer = get_tracer()` and then use that everywhere (this also has perfect parralels we do with logging, eg `logger = logging.getLogger()`)

thinking about the parallel with logging, you set logging with code rather than env vars, eg `logging.basicConfig(format=FORMAT)` 

i guess the argument here would be that there is less configuration complexity needed so its okay to manage it with env vars? still, i think im partial to letting users set it in code - that also makes it possible for users to write their own logging functionality which may be important as we are open source. 

if you think the convienence of global vars is really that high, we could also do both! eg if the user does not set it explicitly, we then check for the global var

agola11 replied:
> if you think the convienence of global vars is really that high, we could also do both! eg if the user does not set it explicitly, we then check for the global var

By global var, do mean environment var?

hwchase17 replied:
oops ya

hwchase17 commented on langchain/tracing/base.py:
```diff
@@ -0,0 +1,85 @@
+        """Start a trace for a tool run."""
+
+    @abstractmethod
+    def end_tool_trace(self, output: str, error: Optional[str] = None) -> None:
```
we still never use this error argument right? if we dont use i would not put in

agola11 replied:
Yeah, so not yet. I have a task for myself to add better error handling in langchain, then pass in the error to the trace where appropriate

hwchase17 replied:
imo would just leave out until then

hwchase17 commented on langchain/tracing/nested_json.py:
```diff
@@ -0,0 +1,147 @@
+
+    _instance = None
+
+    def __new__(cls):
```
are we using __new__ vs the more standard __init__ for a reason? if so, maybe a comment cause its not super typical

agola11 replied:
The idea is that to avoid having to pass the logger through calls in a chain, we need a global shared logger, i.e. a singleton. To prevent the multiple instantiations of this class, we keep track of whether a **class** was instantiated. That's why we use `__new__`... it takes in the class (`cls`) to control object creation where as `__init__`'s job is to take in the instance and handle setting instance variables. 

Good point though, I'll add a comment

hwchase17 commented on langchain/tracing/nested_json.py:
```diff
@@ -0,0 +1,147 @@
+        run = self._stack.pop()
+        if not self._stack:
+            self._execution_order = 1
+            print(run.to_json(indent=2))
```
this just prints it out? wouldnt saving to disk make more sense? otherwise isnt it pretty similar to stdout?

agola11 replied:
Yeah lol great point I'm going to change this to save to disk

hwchase17 commented on langchain/chains/base.py:
```diff
@@ -104,7 +105,9 @@ def __call__(
             )
+        get_tracer().start_chain_trace({"name": self.__class__.__name__}, inputs)
         outputs = self._call(inputs)
+        get_tracer().end_chain_trace(outputs)
```
theres currently the start and end print statements above and below. i would love to make these StdOut tracing functionality equivalent to that

hwchase17 reviewed (commented):
i would just focus on getting the base one right and forget about nested json/stdout for the moment. i dont think we shoudl publicize them until we have a more legit solution anyways so dont see the point of trying to get them in now,

agola11 committed:
fix some lint stuff with singleton instantiation

agola11 committed:
edit dataclasses

agola11 committed:
add langchain tracer

agola11 committed:
format

agola11 committed:
multithreading support

agola11 committed:
refactoring json and langchain tracing

agola11 committed:
Add BaseCallbackHandler and CallbackManager (#476)

agola11 committed:
allow for optional CallbackManager in LLM, Chain, and Agent (#482)

hwchase17 committed:
WIP: stdout callback (#479)

first pass at stdout callback

for the most part, went pretty smoothly. aside from the code here, here
are some comments/observations.

1. 
should somehow default to stdouthandler so i dont have to do 
```
from langchain.callbacks import get_callback_manager
from langchain.callbacks.stdout import StdOutCallbackHandler

get_callback_manager().add_handler(StdOutCallbackHandler())
```

2. I kept around the verbosity flag. 1) this is pretty important for
getting the stdout to look good for agents (and other things). 2) I
actually added this for LLM class since it didn't have it.

3. The only part that isn't basically perfectly moved over is the end of
the agent run. Here's a screenshot of the new stdout tracing
![Screen Shot 2022-12-29 at 4 03 50
PM](https://user-images.githubusercontent.com/11986836/210011538-6a74551a-2e61-437b-98d3-674212dede56.png)

Noticing it is missing logging of the final thought, eg before this is
what it looked like
![Screen Shot 2022-12-29 at 4 13 07
PM](https://user-images.githubusercontent.com/11986836/210011635-de68b3f5-e2b0-4cd3-9f1a-3afe970a8716.png)

The reason its missing is that this was previously logged as part of
agent end (lines 205 and 206)

this is probably only relevant for the std out logger? any thoughts for
how to get it back in?

hwchase17 committed:
Harrison/not verbose (#487)

hwchase17 committed:
add explicit agent end method (#486)

hwchase17 committed:
switch up defaults (#485)

i kinda like this just because we call `self.callback_manager` so many
times, and thats nicer than `self._get_callback_manager()`?

agola11 committed:
restore files

agola11 committed:
Merge remote-tracking branch 'upstream/ankush/callbackhandler' into ankush/base-tracing

agola11 committed:
WIP refactoring

agola11 committed:
finish refactoring

agola11 committed:
update script and notebook

agola11 committed:
quick comment fixes (#494)

hwchase17 committed:
fix verbosity (#496)

1. remove verbose from someplace it didnt relaly belong
2. everywhere else, make verbose Optional[bool] with default to None
3. make base classes accept None, and then look up globabl verbosity if
thats the case

hwchase17 committed:
Harrison/get rid of prints (#490)

deprecate all prints in favor of callback_manager.on_text (open to
better naming)

hwchase17 committed:
remove logger (#491)

remove old logging class (no longer used anyways)

hwchase17 committed:
Harrison/streamlit handler (#488)

also add a set handler method

usage is:
```
from langchain.callbacks.streamlit import StreamlitCallbackHandler
import langchain
langchain.set_handler(StreamlitCallbackHandler())
```

produces the following output


![Screen Shot 2022-12-29 at 10 50 33
PM](https://user-images.githubusercontent.com/11986836/210032762-7f53fffa-cb2f-4dac-af39-7d4cf81e55dd.png)

only works for agent stuff currently

agola11 committed:
toggle flag

agola11 committed:
merge

agola11 committed:
edit script

agola11 committed:
refactor, add fakes for testing

agola11 committed:
unit tests

agola11 committed:
multithreaded test

agola11 committed:
format

agola11 committed:
fix ontext comment (#500)

agola11 committed:
remove all runs before proceeding with test

agola11 committed:
Merge remote-tracking branch 'upstream/ankush/callbackhandler' into ankush/base-tracing

agola11 committed:
rm verbose flag

agola11 committed:
session management for tracing

agola11 committed:
format

agola11 committed:
tests

agola11 committed:
change LLMResult class to be json serializable and run notebook

agola11 committed:
fix session to not have child runs

agola11 committed:
update notebook

agola11 committed:
update notebook

agola11 committed:
add examples for datasets and feedback

agola11 committed:
update notebook

agola11 committed:
add on_error handling

agola11 committed:
merge attempt

agola11 committed:
get tracer updated

agola11 committed:
revert schema

agola11 committed:
fix error handling

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

hwchase17 committed:
cr

agola11 committed:
add docker compose pull, fix docker-compose file

agola11 committed:
fix tracer tests

agola11 committed:
fix tests

agola11 committed:
update tracing schemas (#4)

* fix tracer tests

* update notebook

agola11 committed:
update notebook and docker compose, allow for auth headers in tracer

agola11 committed:
more defensive handling of requests errors

hwchase17 committed:
cr

hwchase17 committed:
default session (#6)

* default session

* cr

agola11 committed:
get tests working

agola11 committed:
refactor tests

agola11 committed:
change notebook, fix loading stuff

agola11 committed:
fix lint and mypy

agola11 committed:
merge

agola11 committed:
whoops

agola11 committed:
fix poetry lock file

agola11 committed:
fix lock

agola11 committed:
inlude on_error to include keyboard interrupt

agola11 committed:
mypy and black

agola11 committed:
flake8

agola11 committed:
add freezegun to test deps

agola11 committed:
cr

agola11 committed:
cr

agola11 committed:
remove tracing notebook

agola11 committed:
update docker-compose
